From bd7d88bb7bdb9003549ab60bc327f62d0cfa3395 Mon Sep 17 00:00:00 2001
From: Fugang Duan <fugang.duan@nxp.com>
Date: Tue, 6 Jun 2017 14:11:24 +0800
Subject: [PATCH 1876/5242] MLK-15031-02 tty: serial: fsl_lpuart: free the rx
 dma buffer when port is closed

commit  a0b2a663b6da08297834273862173fdb6fac9f61 from
https://source.codeaurora.org/external/imx/linux-imx.git

Free the rx dma buffer when the port is closed.

Signed-off-by: Fugang Duan <fugang.duan@nxp.com>
Reviewed-by: Robin Gong <yibin.gong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/tty/serial/fsl_lpuart.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/tty/serial/fsl_lpuart.c b/drivers/tty/serial/fsl_lpuart.c
index 608a5f3..9d748cf 100644
--- a/drivers/tty/serial/fsl_lpuart.c
+++ b/drivers/tty/serial/fsl_lpuart.c
@@ -1050,6 +1050,7 @@ static void lpuart_dma_rx_free(struct uart_port *port)
 	dma_unmap_single(sport->port.dev, sport->dma_rx_buf_bus,
 			sport->rxdma_len, DMA_FROM_DEVICE);
 
+	devm_kfree(sport->port.dev, sport->dma_rx_buf_virt);
 	sport->dma_rx_buf_bus = 0;
 	sport->dma_rx_buf_virt = NULL;
 }
@@ -1488,8 +1489,8 @@ static void lpuart_shutdown(struct uart_port *port)
 			sport->dma_rx_in_progress = false;
 			dmaengine_terminate_all(sport->dma_rx_chan);
 		}
-		lpuart_dma_rx_free(&sport->port);
 		del_timer_sync(&sport->lpuart_timer);
+		lpuart_dma_rx_free(&sport->port);
 	}
 
 	if (sport->lpuart_dma_tx_use) {
@@ -1535,8 +1536,8 @@ static void lpuart32_shutdown(struct uart_port *port)
 			sport->dma_rx_in_progress = false;
 			dmaengine_terminate_all(sport->dma_rx_chan);
 		}
-		lpuart_dma_rx_free(&sport->port);
 		del_timer_sync(&sport->lpuart_timer);
+		lpuart_dma_rx_free(&sport->port);
 	}
 
 	if (sport->lpuart_dma_tx_use) {
-- 
1.7.9.5

